dot/parachain/util/inclusion-emulator/inclusion_emulator.go

```diff
@@ -0,0 +1,766 @@
+	}
+}
+
+// InboundHrmpLimitations constraints on inbound HRMP channels
+type InboundHrmpLimitations struct {
+	ValidWatermarks []uint
+}
+
+// OutboundHrmpChannelLimitations constraints on outbound HRMP channels.
+type OutboundHrmpChannelLimitations struct {
+	BytesRemaining    uint
+	MessagesRemaining uint
+}
+
+// FutureValidationCode represents the future validation code hash, if any, and at what relay-parent
+// number the upgrade would be minimally applied.
+type FutureValidationCode struct {
+	BlockNumber        uint
+	ValidationCodeHash parachaintypes.ValidationCodeHash
+}
+
+// OutboundHrmpChannelModification represents modifications to outbound HRMP channels.
+type OutboundHrmpChannelModification struct {
+	BytesSubmitted    uint
+	MessagesSubmitted uint
+}
+
+// HrmpWatermarkUpdate represents an update to the HRMP Watermark.
+type HrmpWatermarkUpdate struct {
+	Type  HrmpWatermarkUpdateType
+	Block uint
+}
+
+// HrmpWatermarkUpdateType defines the type of HrmpWatermarkUpdate.
+type HrmpWatermarkUpdateType int
+
+const (
+	Head HrmpWatermarkUpdateType = iota
+	Trunk
+)
+
+// Watermark returns the block number of the HRMP Watermark update.
+func (h HrmpWatermarkUpdate) Watermark() uint {
+	return h.Block
+}
+
+// ConstraintModifications represents modifications to constraints as a result of prospective candidates.
+type ConstraintModifications struct {
+	// The required parent head to build upon.
+	RequiredParent *parachaintypes.HeadData
+	// The new HRMP watermark.
+	HrmpWatermark *HrmpWatermarkUpdate
+	// Outbound HRMP channel modifications.
+	OutboundHrmp map[parachaintypes.ParaID]OutboundHrmpChannelModification
+	// The amount of UMP XCM messages sent. `UMPSignal` and separator are excluded.
+	UmpMessagesSent uint
+	// The amount of UMP XCM bytes sent. `UMPSignal` and separator are excluded.
+	UmpBytesSent uint
+	// The amount of DMP messages processed.
+	DmpMessagesProcessed uint
+	// Whether a pending code upgrade has been applied.
+	CodeUpgradeApplied bool
+}
+
+func (cm *ConstraintModifications) Clone() *ConstraintModifications {
+	return &ConstraintModifications{
+		RequiredParent:       cm.RequiredParent,
+		HrmpWatermark:        cm.HrmpWatermark,
+		OutboundHrmp:         maps.Clone(cm.OutboundHrmp),
+		UmpMessagesSent:      cm.UmpMessagesSent,
+		UmpBytesSent:         cm.UmpBytesSent,
+		DmpMessagesProcessed: cm.DmpMessagesProcessed,
+		CodeUpgradeApplied:   cm.CodeUpgradeApplied,
+	}
+}
+
+// Identity returns the 'identity' modifications: these can be applied to
+// any constraints and yield the exact same result.
+func NewConstraintModificationsIdentity() *ConstraintModifications {
+	return &ConstraintModifications{
+		RequiredParent:       nil,
+		HrmpWatermark:        nil,
+		OutboundHrmp:         make(map[parachaintypes.ParaID]OutboundHrmpChannelModification),
+		UmpMessagesSent:      0,
+		UmpBytesSent:         0,
+		DmpMessagesProcessed: 0,
+		CodeUpgradeApplied:   false,
+	}
```

Comment:
These are all init values right
```suggestion
	return &ConstraintModifications{
		OutboundHrmp:         make(map[parachaintypes.ParaID]OutboundHrmpChannelModification),
	}
```